lib/ddtrace/monkey.rb

```diff
@@ -0,0 +1,68 @@
+require 'thread'
+
+# We import all patchers for every module we support, but this is fine
+# because patchers do not include any 3rd party module nor even our
+# patching code, which is required on demand, when patching.
+require 'ddtrace/contrib/elasticsearch/patcher'
+require 'ddtrace/contrib/redis/patcher'
+
+module Datadog
+  # Monkey is used for monkey-patching 3rd party libs.
+  module Monkey
+    @patched = []
+    @autopatch_modules = { elasticsearch: true, redis: true }
+    # Patchers should expose 2 methods:
+    # - patch, which applies our patch if needed. Should be idempotent,
+    #   can be call twice but should just do nothing the second time.
+    # - patched?, which returns true if the module has been succesfully
+    #   patched (patching might have failed if requirements were not here)
+    @patchers = { elasticsearch: Datadog::Contrib::Elasticsearch::Patcher,
+                  redis: Datadog::Contrib::Redis::Patcher }
+    @mutex = Mutex.new
+
+    module_function
+
+    def autopatch_modules
+      @autopatch_modules.clone
+    end
+
+    def patch_all
+      patch @autopatch_modules
+    end
+
+    def autopatch_all
+      patch_all if 'true' == ENV['DATADOG_TRACE_AUTOPATCH'].to_s
```

Comment:
not sure to rely in environment variables to ``autopatch_all``. I would double check the Python client if we're doing something similar.

I prefer to keep separated the "magic" world of a full fledged framework where the autopatch could be something familiar. But don't know, let's use the same approach of Python for now so leave this code as is if Python uses env vars too.